tests/chainer_tests/functions_tests/noise_tests/test_dropout.py

```diff
@@ -129,4 +132,46 @@ def test_immutable(self, backend_config):
         self.check_immutable(self.inputs, backend_config)
 
 
+@testing.parameterize(*testing.product({
+    'specify_mask': [True, False],
+    'train': [True, False],
+}))
+class TestDropoutMask(unittest.TestCase):
+
+    def setUp(self):
+        self.x = numpy.random.uniform(-1, 1, (2, 3)).astype(numpy.float32)
+        self.mask = (numpy.random.uniform(-1, 1, (2, 3)) > 0).astype(
+            numpy.float32)
+
+    def _check(self):
+        mask = self.mask if self.specify_mask else None
+        with chainer.using_config('train', self.train):
+            out, out_mask = functions.dropout(
+                self.x, 0.5, mask=mask, return_mask=True)
+
+        if self.train:
+            assert isinstance(out_mask, type(out.array))
+            if mask is None:
+                assert out_mask.shape == out.array.shape
+            else:
+                assert out_mask is mask
```

Comment:
How about in both cases (mask is None and mask is not None), make sure we get the correct output by applying the returned mask to the input?

Reply:
Test added.